pandas/core/indexes/base.py

```diff
@@ -548,6 +548,9 @@ def _deepcopy_if_needed(self, orig, copy=False):
         """
         if copy:
             # Retrieve the "base objects", i.e. the original memory allocations
+            if not isinstance(orig, np.ndarray):
+                # orig is a DatetimeIndex
+                orig = orig.values
```

Comment:
@jreback This special case is because we sometimes construct DatetimeIndex objects where the `_data` attribute still is a DatetimeIndex in itself. 
To me this seems like something we should not do? (it doesn't hurt, but there is also no reason to have it I think, so better always store the ndarray to be consistent?)

Reply:
Just opened an issue about it: https://github.com/pandas-dev/pandas/issues/20810 
It came up from fixing the warnings in the tests. The above case, where `orig` is not always a ndarray (hence my special case `if` statement), triggered it, because `orig` always comes from a `_data` attribute, and so I expected it to always be a ndarray.